src/WTG.Analyzers/Analyzers/LinqEnumerable/LinqEnumerableAnalyzer.cs

```diff
@@ -0,0 +1,295 @@
+		{
+			if (invocation.ArgumentList.Arguments.Count == 1)
+			{
+				var expression = ((MemberAccessExpressionSyntax)invocation.Expression).Expression.TryGetExpressionFromParenthesizedExpression();
+				return ContainsSingleElement(expression);
+			}
+
+			return ContainsSingleElement(invocation.ArgumentList.Arguments[0].Expression);
+		}
+
+		static bool LooksLikeShouldBeAppend(InvocationExpressionSyntax invocation) => (invocation.ArgumentList.Arguments.Count == 1 ? ContainsSingleElement(invocation.ArgumentList.Arguments[0].Expression) : ContainsSingleElement(invocation.ArgumentList.Arguments[1].Expression));
+
+		static bool LooksLikeCollectionLiteral(ExpressionSyntax e, int? expectedItems = default)
+		{
+			var maybeExpressions = LinqEnumerableUtils.GetInitializer(e)?.Expressions;
+			if (!maybeExpressions.HasValue)
+			{
+				return false;
+			}
+
+			var expressions = maybeExpressions.GetValueOrDefault();
+			if (expectedItems.HasValue && expressions.Count != expectedItems.GetValueOrDefault())
+			{
+				return false;
+			}
+
+			foreach (var expression in expressions)
+			{
+				if (expression.IsKind(SyntaxKind.SimpleAssignmentExpression))
+				{
+					return false;
+				}
+			}
+
+			return true;
+		}
+
+		static bool ContainsSingleElement(ExpressionSyntax e) => LooksLikeCollectionLiteral(e, expectedItems: 1);
+
+		static ExpressionSyntax? CheckConcatExpressionMeetsSemanticRequirements(SemanticModel semanticModel, InvocationExpressionSyntax invocation)
+		{
+			var arguments = invocation.ArgumentList.Arguments;
+
+			var e = ((MemberAccessExpressionSyntax)invocation.Expression).Expression.TryGetExpressionFromParenthesizedExpression();
+
+			var concatMethodSymbol = (IMethodSymbol?)semanticModel.GetSymbolInfo(invocation.Expression).Symbol;
+
+			if (concatMethodSymbol == null || !concatMethodSymbol.IsMatch(WellKnownTypeNames.Enumerable, nameof(Enumerable.Concat)))
+			{
+				return null;
+			}
+
+			switch (arguments.Count)
+			{
+				case 1:
+					if (!(IsSupportedCollection(semanticModel, e) || IsSupportedCollection(semanticModel, arguments[0].Expression)))
+					{
+						return null;
+					}
+
+					break;
+
+				case 2:
+					var typeSymbol = semanticModel.GetTypeInfo(e).Type;
+
+					if (typeSymbol == null || !typeSymbol.IsMatch(WellKnownTypeNames.Enumerable))
+					{
+						return null;
+					}
+
+					if (!AnyArgumentIsSupportedCollection(semanticModel, arguments))
+					{
+						return null;
+					}
+					
+					break;
+
+				default:
+					return null;
+			}
+
+			return invocation.Expression;
+		}
+
+		static bool IsSupportedCollection(SemanticModel model, ExpressionSyntax e)
+		{
+			var typeSymbol = model.GetTypeInfo(e).Type;
+
+			if (typeSymbol == null)
+			{
+				return false;
+			}
+
+			return typeSymbol.TypeKind == TypeKind.Array || typeSymbol.IsMatch(WellKnownTypeNames.List_T);
+		}
+
+		static bool AnyArgumentIsSupportedCollection (SemanticModel semanticModel, SeparatedSyntaxList<ArgumentSyntax> arguments)
```

Comment:
this seems odd (and not just the space before the open parenthesis). do we really only care that 'a' argument is a supported collection? not all the arguments, or at least specific arguments?

Reply:
At the call-site, we already know that there are two arguments. We don't care which of the following it is, as all three will trigger a rule:

```csharp
Enumerable.Concat(new[] { 1 }, foo)
Enumerable.Concat(foo, new[] { 1 })
Enumerable.Concat(new[] { 1 }, new[] { 2 })
```